content/blog/jeststream-eventual-consistency.md

```diff
@@ -0,0 +1,121 @@
++++
+# Delegate with trust
+
+When I speak about async patterns (messaging, event sourcing,…) with other developers.
+They are afraid of eventual consistency.
+It’s not said like that it’s :
+> The frontend needs the answer directly
+> The business process needs to be validated with perfect data
+> My next page needs fresh data to display properly
+
+With messaging system, you have to wait the message to be processed, and the state to be updated, then you can query it with trust. Could be 30ms or more if sh*t appends.
+
+With classing CRUD system, you write on your database, you query it directly with trust.
+
+How can I delegate the job and be sure that it’ll be done or that I get an error if it can’t be done ?
+Let’s dive into use cases and tactics where I can get the same trust using NATS JetStream.
+
+## Their could be only one
+I have a payment system, and in this system we must withdraw the customer for a given orderID only once !
+Sadly our partner who triggers the payment can call us multiple times with sometimes 1 day of delay.
+
+Here we use one subject by order to store its payment log.
+The subject will contains only one message the withdrawal.
+Using publish expectations, I can be sure that there will be only one subject by order’s payment.
+
+```typescript
+// connect to NATS
+const nc = connect();
+// create a jetstream client:
+const js = nc.jetstream();
+const orderID = shortUID();
+// The message should be the first on the subject
+await js.publish(`payment.${orderID}`, Empty, {
+	expect: { lastSubjectSequence: 0 }
+});
+```
+
+## The fat finger
+The customer always clicks multiple times on the same button, so it calls the backend and duplicates messages !
+This triggers side effects multiples times creating spam, and exploding the BI.
+
+NATS JetStream embed a deduplication system based on message ID for a given window. The window is configured in nano seconds at the stream level.
+
+```typescript
+// connect to NATS
+const nc = connect();
+// create the stream
+const jsm = await nc.jetstreamManager();
+await jsm.streams.add({
+	name: "a",
+	subjects: ["a.*"],
+	// Deduplicate for 5 seconds
+	duplicate_window: 5_000_000_000
+});
+// create a jetstream client:
+const js = nc.jetstream();
+// if this message is published twice within 5s, only the first published will be stored
+await js.publish("a.b", Empty, { msgID: "a" });
+```
+
+## Order maintenance
+I have a payment system, for the refund process, I need first to have a double human validation. Then the refund can append. And it must append only once.
+
+For a refund subject, I will say that the refund must be exactly in 3rd place or crash.
+```typescript
+// connect to NATS
+const nc = connect();
+// create a jetstream client:
+const js = nc.jetstream();
+const orderID = shortUID();
+// The message should be the first on the subject
+await js.publish(`refund.${orderID}`, Empty, {
+	expect: { lastSubjectSequence: 3 }
+});
```

Comment:
There are two issues with this example.

If the "third place" is required, the last subject sequence would be 2 since this new message would result in sequence 3.

However, the bigger issue is that the sequence applies to the whole stream, not on a per-subject basis. With a single refund in the stream, this would work, but when the second refund occurs, the sequence of the "third event" would be 6:

```
refund.100 -> seq 1
refund.100 -> seq 2
refund.100 -> seq 3
refund.200 -> seq 4
refund.200 -> seq 5
refund.200 -> seq 6
```

And this assumes these events are not interleaved. A correct way to model this is to either get the last message by subject, assert this is a proper state transition and use that sequence number as the expected one. Of if you need more than just the last message for the subject, you can take the same approach as in the Business rule example below of creating a temporary consumer.